L69-cpp-core-event-manager.md

```diff
@@ -0,0 +1,282 @@
+enum Priority {
+  PRIORITY_MIN = 0,
+  PRIORITY_LOW = 0,
+  PRIORITY_MED = 1,
+  PRIORITY_HIGH = 2,
+  PRIORITY_IMMEDIATE = 3,
+  PRIORITY_SYSTEM = 4,
+  PRIORITY_MAX = 4,
+  NUM_PRIORITIES = 5
+};
+
+// A top-level base class of EventManagerInterface.
+class BaseEventManagerInterface {
+ public:
+  virtual ~BaseEventManagerInterface();
+
+  // Adds a new one-shot alarm, which will be scheduled at the deadline to run the 
+  // closure with true. The deadline here is an absolute time. The AlarmKey pointer 
+  // is an output parameter which can be used to cancel the alarm before it fires. 
+  // Note that the closure should not be blocking or long-running, as it would 
+  // block the worker thread in the EventManager from polling the file 
+  // descriptors. In case that the alarm is cancelled successfully in 
+  // ConsistentCancel, the closure will be run with false in the calling thread of 
+  // ConsistentCancel. Callers of ConsistentCancel need to be cautious to avoid 
+  // deadlock.
+  virtual void ScheduleAlarmIn(gpr_timespec deadline, std::function<void(bool)> c, Priority priority, AlarmKey* key = nullptr) = 0;
+
+  // Adds a closure to the EventManager, which will be run at the earliest possible 
+  // time. The TaskKey pointer is an output parameter which can be used to cancel 
+  // the closure at a later time (currently not supported in 
+  // BaseEventManagerInterface). Note that the closure should not be blocking or 
+  // long-running, as it would block the worker thread in the EventManager from 
+  // polling the file descriptors.
+  virtual void RunFunction(std::function<void()> handler, Priority priority, TaskKey* key = nullptr) = 0;
+
+  struct Task {
+    std::function<void()> handler;
+    Priority priority;
+    TaskKey* key;
+    int internal_added_queue;
+  };
+  // Adds multiple closures to the event management system all at once.
+  virtual void RunFunctions(Task* tasks, int length) = 0;
+
+  // Cancels an alarm which has been added to the EventManager via 
+  // ScheduleAlarmIn(). This function will run the closure with false in the 
+  // calling thread if it successfully cancels the alarm.
+  virtual void ConsistentCancel(const AlarmKey& key) = 0;
+
+  // Cancels an alarm as ConsistentCancel(), but wait if the task is currently 
+  // running.
+  virtual void BlockingConsistentCancel(const AlarmKey& key) = 0;
+
+  // These two methods adjust a refcount of the EventManager, preventing it from 
+  // shutdown if the refcount is positive. ShutdownRef() should be called before 
+  // adding the first closure to the EventManager, and ShutdownUnref() can be 
+  // called after removing the last closure from the EventManager. Also, if you 
+  // need to keep the EventManager serving before receiving an RPC response, you 
+  // need to call ShutdownRef() before sending the RPC, and invoke ShutdownUnref() 
+  // when the response arrives.
+  virtual void ShutdownRef() = 0;
```

Comment:
Has there been any consideration of modeling this lifetime behavior with RAII types? It _is_ a C++ interface, after all. 😉 E.g., some mixture of `std::shared_ptr` and factory functions for creation?

Reply:
I don't think we'll do that for compatibility with the Google EM interface.

Reply:
Can you expand on what compatibility with the Google EM interface means? I'm assuming that there would be (is?) an adapter of some sort between the Google EM and this EM interface if for no other reason than to allow the two to change decoupled from each other. Such an adapter could, for example, turn RAII type construction/destruction into the needed Ref/Unref calls.

It's incredibly useful to have an implementation of an event manager to prove that the architecture works. Thank you for doing that: running code speaks volumes. It just looks like there's a lot of room here to have the type system help with ownership while also being compatible with the Google EM interface.

Reply:
The design is derived from a Google internal EM implementation, and the plan is to have Google EM as one of the implementations of the BaseEventManagerInterface. There are some semantics difference between the two so yes there will be some sort of adapting methods in the Google EM implementation, but as the team just discussed again, though it's a somewhat unideal situation, one of the design consideration is to minimize the differences (hence minimize the number of adapting methods needed) between the API in BaseEventManagerInterface and the existing GoogleEM implementation. So we will to keep the interface similar to the Google EM's interface except for some rare cases (e.g. to enable a big use cases that the current design is not capable of).

As mentioned in the doc, the libuv implementation will be provided and will be using this architecture. It's an work in progress by the team.